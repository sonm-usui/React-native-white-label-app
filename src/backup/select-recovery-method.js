// @flow

import React, { Component } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import type {
  ReactNavigationBackup,
  SelectRecoveryMethodProps,
} from './type-backup'

import {
  selectRecoveryMethodRoute,
  exportBackupFileRoute,
  settingsRoute,
  cloudBackupRoute,
} from '../common'
import { withStatusBar } from '../components/status-bar/status-bar'
import {
  Container,
  CustomView,
  CustomText,
  Icon,
  CustomHeader,
} from '../components'
import { color, colors } from '../common/styles/constant'
import styles from './styles'
import {
  hasVerifiedRecoveryPhrase,
  generateBackupFile,
} from '../backup/backup-store'
import { safeSet, walletSet } from '../services/storage'
import { HAS_VERIFIED_RECOVERY_PHRASE } from './type-backup'
import { appName } from '../external-imports'

const closeImage = require('../images/icon-Close.png')
const backup = require('../images/upload13x.png')
const download = require('../images/download3x.png')

export class SelectRecoveryMethod extends Component<
  SelectRecoveryMethodProps,
  void
> {
  static navigationOptions = ({
    navigation: { navigate },
  }: ReactNavigationBackup) => ({
    header: () => (
      <CustomHeader
        backgroundColor={color.bg.fifth.color}
        largeHeader
        flatHeader
      >
        <CustomView style={[styles.headerSpacer]} />
        <CustomView style={[styles.headerSpacer]}>
          <Icon
            mediumLarge
            onPress={() => navigate(settingsRoute)}
            iconStyle={[styles.headerIcon]}
            src={closeImage}
          />
        </CustomView>
      </CustomHeader>
    ),
    gestureEnabled: false,
    headerShown: true,
  })

  componentDidMount() {
    this.props.hasVerifiedRecoveryPhrase()
    walletSet(HAS_VERIFIED_RECOVERY_PHRASE, 'true')
    safeSet(HAS_VERIFIED_RECOVERY_PHRASE, 'true')
  }

  backup = () => {
    this.props.generateBackupFile()
    const { initialRoute } = this.props.route.params
    this.props.navigation.navigate(exportBackupFileRoute, {
      initialRoute,
    })
  }
  cloudBackup = () => {
    this.props.navigation.navigate(cloudBackupRoute, {
      fromToggleAction: false,
    })
  }
  render() {
    return (
      <Container style={[styles.selectRecoveryMethod]}>
        <CustomView center>
          <CustomText transparentBg center style={[styles.backuptitle]}>
            Select your backup method
          </CustomText>
        </CustomView>
        <Container spaceBetween style={[styles.selectContainer]}>
          <CustomView
            onPress={this.cloudBackup}
            spaceAround
            center
            style={[styles.selectMethod, { backgroundColor: '#86B93B' }]}
          >
            <Icon iconStyle={[{ marginBottom: 10 }]} mediumLarge src={backup} />
            <CustomText center transparentBg style={[styles.title2]}>
              Cloud Backup
            </CustomText>
            <CustomText
              size="14"
              center
              transparentBg
              style={{ color: colors.white }}
            >
              Store an encrypted, anonymous backup of {appName} in the Evernym
              Cloud. You will need your Recovery Phrase and a fresh {appName}
              installation to restore.
            </CustomText>
          </CustomView>
          <CustomView center>
            <CustomText primary transparentBg center style={[styles.title1]}>
              or
            </CustomText>
          </CustomView>
          <CustomView
            onPress={this.backup}
            spaceAround
            secondary
            center
            style={[styles.selectMethod]}
          >
            <Icon
              mediumLarge
              iconStyle={[{ marginBottom: 10 }]}
              src={download}
            />
            <CustomText transparentBg center style={[styles.title2]}>
              Downloaded .zip Backup
            </CustomText>
            <CustomText
              size="14"
              transparentBg
              center
              style={[{ color: colors.white }]}
            >
              Manually choose where to store your backup file. You will need
              your Recovery Phrase, a fresh install of {appName} and your
              backup .zip file on the device you wish to restore to.
            </CustomText>
          </CustomView>
        </Container>
      </Container>
    )
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      hasVerifiedRecoveryPhrase,
      generateBackupFile,
    },
    dispatch
  )

export const selectRecoveryMethodScreen = {
  routeName: selectRecoveryMethodRoute,
  screen: withStatusBar({ color: colors.white })(
    connect(null, mapDispatchToProps)(SelectRecoveryMethod)
  ),
}
