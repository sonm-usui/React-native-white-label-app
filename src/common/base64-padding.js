// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
