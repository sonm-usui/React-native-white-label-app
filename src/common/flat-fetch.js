// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
