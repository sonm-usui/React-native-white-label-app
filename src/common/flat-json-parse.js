// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
