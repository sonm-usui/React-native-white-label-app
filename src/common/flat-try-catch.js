// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
