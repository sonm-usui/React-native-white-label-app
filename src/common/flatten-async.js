// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
