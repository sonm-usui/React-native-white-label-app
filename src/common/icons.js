// @flow
import React from 'react'
import { Icon } from 'react-native-eva-icons'
import { colors } from './styles/constant'
import { moderateScale } from 'react-native-size-matters'

type IconProps = {
  width?: number,
  height?: number,
  color?: string,
}

type CommonIconProps = {
  name: string,
} & IconProps

export const HOME_ICON = 'home-outline'
export const CONNECTIONS_ICON = 'people-outline'
export const SETTINGS_ICON = 'settings-2-outline'

export const ERROR_ICON = 'alert-circle-outline'
export const CAMERA_ICON = 'camera-outline'
export const CHECK_MARK_ICON = 'checkmark-circle-2-outline'
export const HOME_MENU_ICON = 'menu-outline'
export const ANDROID_BACK_ARROW_ICON = 'arrow-back-outline'
export const IOS_BACK_ARROW_ICON = 'arrow-ios-back-outline'
export const MORE_ICON = 'more-vertical-outline'
export const DELETE_ICON = 'trash-2-outline'
export const CLOSE_ICON = 'close-outline'
export const CHAT_ICON = 'message-square-outline'
export const INFO_ICON = 'info-outline'
export const ARROW_RIGHT_ICON = 'chevron-right-outline'
export const LOCK_ICON = 'lock'
export const SAVE_ICON = 'save-outline'
export const ALERT_ICON = 'alert-circle-outline'
export const ARROW_FORWARD_ICON = 'arrow-ios-forward-outline'
export const CHECKMARK_ICON = 'checkmark-outline'
export const SHARE_ICON = 'share-outline'
export const GRID_ICON = 'grid-outline'

// common icon class is implemented to set default values (except name) for icons
// in that case we should only set icon name and get an icon with default color and size
export const EvaIcon = (props: CommonIconProps) => {
  const { width, height, color } = props

  return (
    <Icon
      {...props}
      width={width ? width : moderateScale(24)}
      height={height ? height : moderateScale(24)}
      fill={color ? color : colors.gray2}
    />
  )
}
