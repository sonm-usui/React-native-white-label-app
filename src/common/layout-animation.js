// @flow

import { LayoutAnimation } from 'react-native'
import { checkIfAnimationToUse } from '../bridge/react-native-cxs/RNCxs'

export function animateLayout() {
  if (!checkIfAnimationToUse()) {
    LayoutAnimation.configureNext(
      LayoutAnimation.create(
        150,
        LayoutAnimation.Types.linear,
        LayoutAnimation.Properties.opacity
      )
    )
  }
}
