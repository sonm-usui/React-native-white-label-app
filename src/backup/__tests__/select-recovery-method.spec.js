// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import { SelectRecoveryMethod } from '../select-recovery-method'
import { selectRecoveryMethodRoute } from '../../common'

describe('<SelectRecoveryMethod />', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
  })

  const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    setParams: jest.fn(),
  }
  const route = {
    params: {
      initialRoute: selectRecoveryMethodRoute,
      navigateBack: jest.fn(),
    },
  }

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <SelectRecoveryMethod
          navigation={navigation}
          route={route}
          hydrateCloudBackup={jest.fn()}
          hasVerifiedRecoveryPhrase={jest.fn()}
          generateBackupFile={jest.fn()}
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
