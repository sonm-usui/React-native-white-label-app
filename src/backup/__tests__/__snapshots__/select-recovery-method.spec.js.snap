// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectRecoveryMethod /> should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "flex-end",
      "width": "100%",
    }
  }
>
  <View
    center={true}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "fontFamily": "System",
          },
          Object {
            "backgroundColor": "#333333",
            "color": "#ffffff",
          },
          null,
          Object {
            "textAlign": "center",
          },
          Object {
            "backgroundColor": "transparent",
          },
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          null,
          Object {
            "color": "#505050",
            "fontSize": 20,
            "fontWeight": "400",
            "lineHeight": 22,
            "marginBottom": 10,
            "width": "100%",
          },
        ]
      }
      suppressHighlighting={true}
    >
      Select your backup method
    </Text>
  </View>
  <View
    spaceBetween={true}
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "maxHeight": 310,
      }
    }
  >
    <View
      accessible={true}
      center={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 20,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      spaceAround={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#86B93B",
          "borderRadius": 4,
          "justifyContent": "space-around",
          "padding": 20,
        }
      }
    >
      <View
        clearBg={true}
        iconStyle={
          Array [
            Object {
              "marginBottom": 10,
            },
          ]
        }
        mediumLarge={true}
        src={
          Object {
            "default": "RelativeImageStub",
          }
        }
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <View
          style={Object {}}
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "default": "RelativeImageStub",
              }
            }
            style={
              Array [
                Object {
                  "height": 40,
                  "width": 40,
                },
                null,
                Object {
                  "marginBottom": 10,
                },
              ]
            }
          />
        </View>
      </View>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "color": "#FFFFFF",
              "fontSize": 22,
              "fontWeight": "400",
              "lineHeight": 22,
              "marginBottom": 10,
              "width": "100%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Cloud Backup
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            undefined,
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
          ]
        }
        suppressHighlighting={true}
      >
        Store an encrypted, anonymous backup of 
        AppName
         in the Evernym Cloud. You will need your Recovery Phrase and a fresh 
        AppName
        installation to restore.
      </Text>
    </View>
    <View
      center={true}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            Object {
              "color": "#333333",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "fontSize": 20,
              "fontWeight": "400",
              "lineHeight": 22,
              "marginBottom": 10,
              "width": "100%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        or
      </Text>
    </View>
    <View
      accessible={true}
      center={true}
      focusable={true}
      hitSlop={
        Object {
          "bottom": 20,
          "left": 20,
          "right": 20,
          "top": 20,
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      secondary={true}
      spaceAround={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#777777",
          "borderRadius": 4,
          "justifyContent": "space-around",
          "padding": 20,
        }
      }
    >
      <View
        clearBg={true}
        iconStyle={
          Array [
            Object {
              "marginBottom": 10,
            },
          ]
        }
        mediumLarge={true}
        src={
          Object {
            "default": "RelativeImageStub",
          }
        }
        style={
          Object {
            "backgroundColor": "transparent",
          }
        }
      >
        <View
          style={Object {}}
        >
          <Image
            resizeMode="contain"
            source={
              Object {
                "default": "RelativeImageStub",
              }
            }
            style={
              Array [
                Object {
                  "height": 40,
                  "width": 40,
                },
                null,
                Object {
                  "marginBottom": 10,
                },
              ]
            }
          />
        </View>
      </View>
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "color": "#FFFFFF",
              "fontSize": 22,
              "fontWeight": "400",
              "lineHeight": 22,
              "marginBottom": 10,
              "width": "100%",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Downloaded .zip Backup
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
            undefined,
            Object {
              "fontFamily": "System",
            },
            Object {
              "backgroundColor": "#333333",
              "color": "#ffffff",
            },
            null,
            Object {
              "textAlign": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            Object {
              "color": "#FFFFFF",
            },
          ]
        }
        suppressHighlighting={true}
      >
        Manually choose where to store your backup file. You will need your Recovery Phrase, a fresh install of 
        AppName
         and your backup .zip file on the device you wish to restore to.
      </Text>
    </View>
  </View>
</View>
`;
